pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,342 @@
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type DrainOptions struct {
+	client             *client.Client
+	factory            *cmdutil.Factory
+	Force              bool
+	GracePeriodSeconds int
+	mapper             meta.RESTMapper
+	nodeInfo           *resource.Info
+	out                io.Writer
+	typer              runtime.ObjectTyper
+}
+
+const (
+	lame_long = `Mark nodes as unschedulable.
+`
+	lame_example = `# Mark node "foo" as unschedulable.
+$ kubectl lame foo
+`
+)
+
+func NewCmdLame(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	return &cobra.Command{
+		Use:     "lame NODE",
+		Short:   "Mark node as unschedulable",
+		Long:    lame_long,
+		Example: lame_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunLame(true))
+		},
+	}
+}
+
+const (
+	unlame_long = `Mark nodes as schedulable.
+`
+	unlame_example = `# Mark node "foo" as schedulable.
+$ kubectl unlame foo
+`
+)
+
+func NewCmdUnlame(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	return &cobra.Command{
+		Use:     "unlame NODE",
+		Short:   "Mark node as schedulable",
+		Long:    unlame_long,
+		Example: unlame_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunLame(false))
+		},
+	}
+}
+
+const (
+	drain_long = `Drain node in preparation for maintenance.
+
+The given node will be marked unschedulable to prevent new pods from arriving.
+Then all running pods will be removed from the machine.  If there are any
+unreplicated pods, no pods will be touched without --force.
+
+When you are ready to put the node back into service, use kubectl unlame.
+`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunDrain(cmd))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
+	return cmd
+}
+
+// SetupDrain populates some fields from the factory, grabs command line
+// arguments and looks up the node using Builder
+func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
```

Comment:
Michail, please open a PR to the code conventions doc that describes the
options pattern for kubectl command.

On Nov 14, 2015, at 9:35 AM, Michail Kargakis <notifications@github.com>
wrote:

In pkg/kubectl/cmd/drain.go
<https://github.com/kubernetes/kubernetes/pull/16698#discussion_r44859088>:

> +		Short:   "Drain nodes in preparation for maintenance",
> +		Long:    drain_long,
> +		Example: drain_example,
> +		Run: func(cmd *cobra.Command, args []string) {
> +			cmdutil.CheckErr(options.SetupDrain(cmd, args))
> +			cmdutil.CheckErr(options.RunDrain(cmd))
> +		},
> +	}
> +	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
> +	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
> +	return cmd
> +}
> +
> +// SetupDrain populates some fields from the factory, grabs command line
> +// arguments and looks up the node using Builder
> +func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {

Mind renaming this to Complete? We follow this pattern downstream
<https://github.com/openshift/origin/blob/master/docs/cli_hacking_guide.adoc#command-structure>
and would like to see it used in kubectl too.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16698/files#r44859088>.